packages/cli/src/cog/builder.ts

```diff
@@ -120,34 +117,31 @@ export class CogBuilder {
         const polygons = await Promise.all(coordinates);
 
         if (projection == null) throw new Error('No projection detected');
-        if (resolution == -1) throw new Error('No resolution detected');
+        if (resX == -1) throw new Error('No resolution detected');
         if (bands == -1) throw new Error('No image bands detected');
 
         return {
             projection: projection.code,
             nodata,
             bands,
             bounds: GeoJson.toFeatureCollection(polygons),
-            resolution,
+            pixelScale: resX,
+            resZoom: this.getTiffResZoom(resX),
         };
     }
 
     /**
      * Find the closest resolution to the tiff image
      * @param tiff
      */
-    getTiffResolution(tiff: CogTiff): number {
-        const image = tiff.getImage(0);
-
+    getTiffResZoom(resX: number): number {
         // Get best image resolution
-        const [resX] = image.resolution;
-        let z = 30;
-        while (z > 0) {
-            const currentZoom = InitialResolution / (1 << z);
-            if (currentZoom >= resX) {
-                break;
+        let z = -1;
+        for (const zoom of this.targetProj.tms.zooms) {
+            z = parseInt(zoom.identifier);
```

Comment:
the index is the zoom, Im not sure we want `zoom.identifier` knowledge here, 

```
for (let z  =0; z < zooms.length...)

```

should this also throw if it cannot find one?

Reply:
zooms is a sparse array so `for(;i < len;++i)` not appropriate with sparse array

Reply:
it shouldnt be a sparse array? what TMS has it as sparse?

Reply:
line 36 `tile.matrix.set.ts`

```typescript
        for (const z of def.tileMatrix) {
            const zoomIndex = parseInt(z.identifier);
            if (this.zooms[zoomIndex]) throw new Error(`Duplicate tileMatrix identifier ${z.identifier}`);
            if (z.tileHeight != z.tileWidth) throw new Error('Only square tiles supported');
            if (z.tileHeight != this.tileSize) throw new Error('All tiles must have the same tile size');
            this.zooms[parseInt(z.identifier)] = z;
        }
```

Also not the redundant parseInt instead of using `zoomIndex`

Reply:
This actually sets all the values between 0-X, i am not sure it is a vaild TMS unless it has a definition for all zoom levels.

Reply:
also unsure if we should really be `parseInt` it in the first place, might be a idea to convert this to a `for( let i =0; ...`

or even a `this.zooms = def.tileMatrix.map(c => ...`

Reply:
If we use `map` we need to ensure that the id matches the index. Shall I change the code to do that?

Reply:
This is my uncertainty, I do not know if that is part of the spec. for the TMS we currently have the id is a string that is the zoom index

Reply:
Looking at the spec, this identifier is a `ows:CodeType` which doesnt seem to correlate to a zoom index :shrug:

Reply:
Now sorts the zooms by `scaleDenominator` and maps to `this.zooms`